File: Problem 115.hs
import Data.Function.Memoize (memoFix2)

fill_count :: Integer -> Integer
fill_count = memoFix2 tiles 50

tiles f m n
	| n < m = 1
	| otherwise = sum $ map (\i -> f m (if i < n then n-i-1 else n-i)) (0:[m..n])

main = print $ until ((>=10^6) . fill_count) (+1) 50

File: Problem 115.py
memoize = {}
def tiles(n):
    if n < 50:
        return 1
    if n not in memoize:
        answer = tiles(n-1)
        for i in range(50, n+1):
            if i < n:
                answer += tiles(n-i-1)
            else:
                answer += tiles(n-i)
        memoize[n] = answer
    return memoize[n]

answer = 1
while tiles(answer) < 10**6:
    answer += 1
print(answer)
